lib/dolly/mango/selector.rb

```diff
@@ -0,0 +1,107 @@
+module Dolly
+  module Mango
+    module Selector
+
+      SELECTOR     = 'selector'.freeze
+      EQ_OPERATOR  = '$eq'.freeze
+      NE_OPERATOR  = '$ne'.freeze
+      IN_OPERATOR  = '$in'.freeze
+      GT_OPERATOR  = '$gt'.freeze
+      LT_OPERATOR  = '$lt'.freeze
+      OR_OPERATOR  = '$or'.freeze
+      NOT_SELECTOR = '$not'.freeze
+      NOR_SELECTOR = '$nor'.freeze
+      AND_OPERATOR = '$and'.freeze
+      GTE_OPERATOR = '$gte'.freeze
+      LTE_OPERATOR = '$lte'.freeze
+      EM_OPERATOR  = '$elemMatch'.freeze
+
+      def select_operator_map
+        {
+          eq:            ->(name, value) { build_equal_selector name, value },
+          ne:            ->(name, value) { build_not_equal_selector name, value },
+          in:            ->(name, value) { build_inclusion_selector name, value },
+          not:           ->(name, value) { build_not_selector name, value },
+          nor:           ->(name, value) { build_nor_selector name, value},
+          gt:            ->(name, value) { build_gt_selector name, value },
+          gte:           ->(name, value) { build_gte_selector name, value },
+          lt:            ->(name, value) { build_lt_selector name, value },
+          lte:           ->(name, value) { build_lte_selector name, value },
+          em:            ->(name, value) { build_element_match_selector name, value },
+          [:em, :gt] =>  ->(name, value) { build_element_match_gt_selector name, value },
+          [:em, :gte] => ->(name, value) { build_element_match_gte_selector name, value },
+          [:em, :lt] =>  ->(name, value) { build_element_match_lt_selector name, value },
+          [:em, :lte] => ->(name, value) { build_element_match_lte_selector name, value },
+          [:em, :or] =>  ->(name, value) { build_element_match_or_selector name, value },
+          [:em, :and] => ->(name, value) { build_element_match_and_selector name, value }
+        }.freeze
+      end
+
+      private
+
+      def build_equal_selector name, value
```

Comment:
I disagree. This would create a larger scope api in document model.  

Consider for a monent that if that is desired, this is how the scopes would Look in the Document class.

    mango_scope :not_this_year, ->(year) { build_excluding_selector('year', year, Dolly::Mango::Selector::NOT_OPERATOR) }

part of the reason to have definite methods for each type of selector (as private methods in the selector module) was to simplify the api on the class which you intend to add mango_scopes to.  

I am of the opinion that having one method that accepts the proper operators (which are mapped in the underlying implementation) is a simpler approach for developers than having a lot of methods that more or less all build the selector.  

> I think using a single method with named parameters can keep to be PIE and achieving more a DRY approach

You also brought up using named parameters, However I can't see an example of them in the psuedo code.  A bit confused on that part of your comment.